iml-agent/src/daemon_plugins/ntp.rs

```diff
@@ -0,0 +1,340 @@
+>(
+    is_ntp_configured_by_iml: fn() -> F1,
+    get_time_sync_services: fn() -> F2,
+    ntpd_synced: fn() -> F3,
+    chronyd_synced: fn() -> F4,
+) -> Result<Output, ImlAgentError> {
+    let configured = is_ntp_configured_by_iml().await?;
+
+    tracing::debug!("Configured: {:?}", configured);
+
+    let (ntp_runstate, chrony_runstate) = get_time_sync_services().await?;
+
+    tracing::debug!(
+        "ntp runstate: {:?}. chrony runstate: {:?}",
+        ntp_runstate,
+        chrony_runstate
+    );
+
+    let state = if ntp_runstate >= RunState::Started && chrony_runstate >= RunState::Started {
+        // Both chronyd and ntpd should not be running at the same time.
+        time::State::Multiple
+    } else if ntp_runstate < RunState::Started && chrony_runstate < RunState::Started {
+        // Neither chronyd or ntpd are running
+        time::State::None
+    } else if ntp_runstate >= RunState::Started {
+        let (ntp_synced, ntp_offset) = ntpd_synced().await?;
+
+        match ntp_synced {
+            Some(time::Synced::Synced) => time::State::Synced,
+            Some(time::Synced::Unsynced) => time::State::Unsynced(ntp_offset),
+            None => time::State::Unknown,
+        }
+    } else {
+        let (chrony_synced, chrony_offset) = chronyd_synced().await?;
+
+        match chrony_synced {
+            Some(time::Synced::Synced) => time::State::Synced,
+            Some(time::Synced::Unsynced) => time::State::Unsynced(chrony_offset),
+            None => time::State::Unknown,
+        }
+    };
+
+    let x = serde_json::to_value(state).map(Some)?;
+
+    Ok(x)
+}
+
+impl DaemonPlugin for Ntp {
+    fn start_session(
+        &mut self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        self.update_session()
+    }
+
+    fn update_session(
+        &self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        Box::pin(time_synced(
+            is_ntp_configured_by_iml,
+            get_time_sync_services,
+            ntpd_synced,
+            chronyd_synced,
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use futures::FutureExt;
+
+    #[test]
+    fn test_is_ntp_synced() {
+        let s = r#"b true"#;
+
+        assert_eq!(parse_ntp_synced(s), Some(time::Synced::Synced));
+    }
+
+    #[test]
+    fn test_is_ntp_not_synced() {
+        let s = r#"b false"#;
+
+        assert_eq!(parse_ntp_synced(s), Some(time::Synced::Unsynced));
+    }
+
+    #[test]
+    fn test_bad_output() {
+        let s = r#"b tr"#;
+
+        assert_eq!(parse_ntp_synced(s), None);
+    }
+
+    #[test]
+    fn test_get_ntp_time_offset() {
+        let s = r#"synchronised to NTP server (10.73.10.10) at stratum 12
+time correct to within 949 ms
+polling server every 64 s"#;
+
+        assert_eq!(parse_ntp_time_offset(s), Some("949 ms".to_string().into()));
```

Comment:
I think it would be better to have something more explicit, like `Some(949)`. But I could not find `time::Offset` :)

Reply:
We could probably represent it as a duration + a Sign (postitive, negative). However, we don't actually use this value anywhere on the manager side.